R/statistics.R

```diff
@@ -691,3 +691,247 @@ SecretarySampler <- R6::R6Class("SecretarySampler", public = list(
+MultiArmedBandit <- R6::R6Class(
+    "MultiArmedBandit",
+    public = list(
+        #' @description Creates a new \code{MultiArmedBandit} streamer object.
+        #'
+        #' @param K the number of arms
+        #' @param T number of rounds / draws
+        #'
+        #' @return The new \code{MultiArmedBandit} (invisibly)
+        initialize = function(K, T, N = NULL) {
+            private$K <- K
+            private$T <- T
+            arm_info <- list()
+            for (k in seq(K)) {
+                arm_info[[k]] <- list(mean = CMA$new())
+            }
+            private$arm_info <- arm_info
+            invisible(self)
+        },
+        #' @description Update the \code{MultiArmedBandit} streamer object.
+        #'
+        #' @param x a single draw from an arm
+        #' @param k index of the chosen arm
+        #'
+        #'
+        #' @return The updated code{MultiArmedBandit} (invisibly)
+        update = function(x, k) {
+            if (length(x) > 1) {
+                stop("Only single-value updated allowed")
+            }
+            private$arm_info[[k]]$mean$update(x)
+            private$n_observed <- private$n_observed + 1
+        }
+    ),
+    active = list(
+        #' @description Returns the the list containing values known to
+        #' to the Bandit in the current state.
+        #'
+        #'
+        #' @return list with summary of the state of the Bandit.
+        value = function() {
+            private$get_state()
+        }
+    ),
+    private = list(
+        K = NULL,
+        T = NULL,
+        N = NULL,
+        which = NULL,
+        n_observed = 0,
+        state = NULL,
+        arm_info = NULL,
+        #' @description Creates and returns a list with information about
+        #' the values known to the Bandit in the current state.
+        #'
+        #' @return list with summary of the state of the Bandit.
+        get_state = function() {
+            arm_means <- c()
+            for (k in seq(K)) {
+                arm_means[k] <- private$arm_info[[k]]$mean$value
+            }
+            state <- list(
+                which = private$which,
+                state = private$state,
+                n_observed = private$n_observed,
+                arm_means = arm_means
+            )
+            state
+        }
+    )
+)
+#' Create a streamer for the uniform exploration strategy in the multi-armed
+#' bandit problem.
+#'
+#' @description \code{ExploreFristNBandit}
+#'
+#' @docType class
+#'
+#' @examples
+#' means <- c(1, 2)
+#' N <- 3
+#' T <- 10
+#' K <- length(means)
+#' Bandit <- ExploreFristNBandit$new(K, T, N)
+#' for (t in seq(T)) {
+#'     # get which arm to choose
+#'     k <- Bandit$value$which
+#'     # update the streamer with a value drawn from the kth arm
+#'     Bandit$update(rnorm(1, mean = means[k]), k)
+#' }
+#'
+#' @export
+#' @format An \code{\link{R6Class}} generator object
+ExploreFristNBandit <- R6::R6Class(
```

Comment:
Typo + this is typically called epsilon-first

Reply:
Again, I relied on the _Introduction to Multi-Armed Bandits_ by Aleksandrs Slivkin, where they call it an Explore-First algorithm with a parameter N.

Reply:
Okay, perhaps just `ExploreFirst` then. I think the `N` is evident from the init function.